src/components/legend/draw.js

```diff
@@ -47,365 +47,415 @@ module.exports = function draw(gd) {
     if(typeof gd.firstRender === 'undefined') gd.firstRender = true;
     else if(gd.firstRender) gd.firstRender = false;
 
-    var legend = fullLayout._infolayer.selectAll('g.legend')
-        .data([0]);
+    var legend = createRootNode(),     // root element (sets legend position)
+        clipPath = createClipPath(),   // legend clip path
+        bg = createBackground(),       // legend background and border
+        scrollBox = createScrollBox(), // contains all the legend traces
+        scrollBar = createScrollBar(), // scroll bar (visible only if needed)
+        groups = createGroups(),       // legend groups
+        traces = createLegendTraces(); // legend traces
```

Comment:
I'm not a fan of this pattern.

I'm more in the school of _don't make a bunch of 10-line functions if you're only going to use them once_.

@mdtusz this is mostly your code, what are your thoughts on this?